Migrate `zeek-archiver` into this repo

ckreibich opened the pull request:
This PR does not contain 85 commits, at least not 85 new ones! :-) It moves the `zeek-archiver` master into this repository, inheriting the full commit history. It then adds a handful of commits to unify the content of the original branch with how we're doing things in `zeek-aux`.

The motivation for this is to reduce the number of places we need to keep in sync when doing CI maintenance updates. The reasons for keeping `zeek-archiver` in a separate repo seem largely historic at this point, and it's analogous to `zeek-cut` in many ways, hence this move.

The git magic to pull in the other repo's master is in a84802bd21ba85010b9eb8e637e86497418ad977. The actual new bits are in b842d16edd4abd5ecc6f23579b29d2186d771e5d and after.

Assuming we're all okay with this, it likely needs a quick check from you @0xxon to make sure it doesn't screw up your binary builds. Upon merge we could then update Zeek to ditch the `zeek-archiver` submodule and archive its repo.

CodeQL triggers one warning that is wrong, pretty much as usual — I'm thinking we should remove CodeQL here too.

bbannier commented on zeek-archiver/COPYING.pre-zeek-aux:
@0xxon might have clearer guidelines, but my gut feeling is that the `zeek-archiver` subdirectory needs to keep this license, i.e., this file should stay `COPYING` or `LICENSE`, and the top-level directory should probably reference the deviating license of the subdirectory in e.g., a new file.

ckreibich replied:
The toplevel `COPYING` already says "some files in the distribution may carry their own copyright notices", and since this is us moving our own code I think it's pretty minor. It's also odd that zeek-archiver's license said Corelight. One thought though: assuming we keep `COPYING.pre-zeek-aux`, we should put an ending year on it, since it currently just says "2020".

timwoj replied:
We discussed the licensing stuff in a call the other day and the consensus was to leave the Corelight license in place, given the top level has a mention of third-party licenses.

bbannier commented on zeek-archiver/CHANGES.pre-zeek-aux:
Also, let's make sure to mark https://github.com/ckreibich/zeek-archiver as archived when this gets merged.

ckreibich commented:
Does this look okay to folks now? I've renamed zeek-archiver's license file back to `COPYING`, and added a final year in it (so it says 2020-2023 now).

bbannier commented on zeek-archiver/COPYING:
```diff
@@ -0,0 +1,28 @@
+Copyright (c) 2020 - 2023 by Corelight, Inc
```
```suggestion
Copyright (c) 2020-2023 by Corelight, Inc
```

bbannier commented on testing/zeek-archiver/disable-compression.test:
```diff
@@ -0,0 +1,12 @@
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v --compress="" $(queue_dir) $(archive_dir)
+
+test "$(cat $(archive_date_dir)/${log_out})" == "hello"
```
```suggestion
test "$(cat "$(archive_date_dir)/${log_out}")" == "hello"
```

bbannier commented on testing/zeek-archiver/failing-compress.test:
```diff
@@ -0,0 +1,15 @@
+# (or removed) when the compression command fails.
+# @TEST-EXEC: bash %INPUT
+
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/failing-compress.test:
```diff
@@ -0,0 +1,15 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+log_in=test__2020-07-16-09-43-10__2020-07-16-09-43-10__.log
+log_out=test.09\:43\:10-09\:43\:10.log
```
See above.

```suggestion
log_out=test.09:43:10-09:43:10.log
```

bbannier commented on testing/zeek-archiver/failing-compress.test:
```diff
@@ -0,0 +1,15 @@
+log_out=test.09\:43\:10-09\:43\:10.log
+
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v --compress=false,/bin/false $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v --compress=false,/bin/false "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/metadata-addl.test:
```diff
@@ -0,0 +1,12 @@
+# Verify log name metadata behavior: check that additional pid metadata gets dropped.
+# @TEST-EXEC: bash %INPUT
+
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/metadata-addl.test:
```diff
@@ -0,0 +1,12 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+log_in=test__2020-07-16-09-43-10__2020-07-16-09-43-10__log_suffix=logger-test,pid=4711__.log
+log_out=test.09\:43\:10-09\:43\:10-logger-test.log
```
See above.

```suggestion
log_out=test.09:43:10-09:43:10-logger-test.log
```

bbannier commented on testing/zeek-archiver/metadata-addl.test:
```diff
@@ -0,0 +1,12 @@
+log_out=test.09\:43\:10-09\:43\:10-logger-test.log
+
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/metadata-addl.test:
```diff
@@ -0,0 +1,12 @@
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
+
+test "$(gunzip <$(archive_date_dir)/${log_out}.gz)" == "hello"
```
```suggestion
test "$(gunzip <"$(archive_date_dir)/${log_out}.gz")" == "hello"
```

bbannier commented on testing/zeek-archiver/metadata-empty.test:
```diff
@@ -0,0 +1,12 @@
+# Verify log name metadata behavior: empty metadata is acceptable.
+# @TEST-EXEC: bash %INPUT
+
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/metadata-empty.test:
```diff
@@ -0,0 +1,12 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+log_in=test__2020-07-16-09-43-10__2020-07-16-09-43-10____.log
+log_out=test.09\:43\:10-09\:43\:10.log
```
```suggestion
log_out=test.09:43:10-09:43:10.log
```

bbannier commented on testing/zeek-archiver/metadata-empty.test:
```diff
@@ -0,0 +1,12 @@
+log_out=test.09\:43\:10-09\:43\:10.log
+
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/metadata-empty.test:
```diff
@@ -0,0 +1,12 @@
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
+
+test "$(gunzip <$(archive_date_dir)/${log_out}.gz)" == "hello"
```
```suggestion
test "$(gunzip <"$(archive_date_dir)/${log_out}.gz")" == "hello"
```

bbannier commented on testing/zeek-archiver/metadata-invalid.test:
```diff
@@ -0,0 +1,21 @@
+# @TEST-EXEC: bash %INPUT
+
+@TEST-START-FILE run.sh
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/metadata-invalid.test:
```diff
@@ -0,0 +1,21 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/metadata.test:
```diff
@@ -0,0 +1,12 @@
+# Verify log name metadata behavior: check that suffixes get applied correctly.
+# @TEST-EXEC: bash %INPUT
+
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/metadata.test:
```diff
@@ -0,0 +1,12 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+log_in=test__2020-07-16-09-43-10__2020-07-16-09-43-10__log_suffix=logger-test__.log
+log_out=test.09\:43\:10-09\:43\:10-logger-test.log
```
See above.

```suggestion
log_out=test.09:43:10-09:43:10-logger-test.log
```

bbannier commented on testing/zeek-archiver/metadata.test:
```diff
@@ -0,0 +1,12 @@
+log_out=test.09\:43\:10-09\:43\:10-logger-test.log
+
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/metadata.test:
```diff
@@ -0,0 +1,12 @@
+echo hello > "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
+
+test "$(gunzip <$(archive_date_dir)/${log_out}.gz)" == "hello"
```
```suggestion
test "$(gunzip <"$(archive_date_dir)/${log_out}.gz")" == "hello"
```

bbannier commented on testing/zeek-archiver/pre-compression-logs.test:
```diff
@@ -0,0 +1,13 @@
+# Detect pre-compressed logs.
+# @TEST-EXEC: bash %INPUT
+
+. $SCRIPTS/zeek-archiver-common.sh
```
```suggestion
. "$SCRIPTS/zeek-archiver-common.sh"
```

bbannier commented on testing/zeek-archiver/pre-compression-logs.test:
```diff
@@ -0,0 +1,13 @@
+. $SCRIPTS/zeek-archiver-common.sh
+
+log_in=test__2020-07-16-09-43-10__2020-07-16-09-43-10__.log
+log_out=test.09\:43\:10-09\:43\:10.log
```
```suggestion
log_out=test.09:43:10-09:43:10.log
```

bbannier commented on testing/zeek-archiver/pre-compression-logs.test:
```diff
@@ -0,0 +1,13 @@
+
+echo hello > "$(queue_dir)/${log_in}"
+gzip "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
```
```suggestion
zeek-archiver -1 -v "$(queue_dir)" "$(archive_dir)"
```

bbannier commented on testing/zeek-archiver/pre-compression-logs.test:
```diff
@@ -0,0 +1,13 @@
+gzip "$(queue_dir)/${log_in}"
+zeek-archiver -1 -v $(queue_dir) $(archive_dir)
+
+test "$(gunzip <$(archive_date_dir)/${log_out}.gz)" == "hello"
```
```suggestion
test "$(gunzip <"$(archive_date_dir)/${log_out}.gz")" == "hello"
```

bbannier commented on testing/zeek-archiver/command-injection-filenames.test:
All issues I flagged here can be detected automatically by a future pre-commit hook with https://www.shellcheck.net/ if the files are identifiable as shell scripts. This is currently broken by the files having the redundant `.test` extension (they are already in the `testing/` folder), could you instead please use e.g., `.sh` as extension?

bbannier replied:
True, I jumped onto it once you dragged them into the light 😬

ckreibich replied:
LOL, I swear this wasn't intentional nerd-sniping. :-)

timwoj replied:
That commit message though 😛

bbannier reviewed (approved):
Thanks for looking into `LICENSE`, this makes sense to me. 🚀 

I made a pass over the tests to make sure they work with unusual paths (maybe not so unusual?). At some point we should really have a pre-commit setup to move such stuff upfront. This requires that files have sensible extensions, so it would help a lot if you could rename the `.test` files to `.sh` now.
